### PR TITLE
Add `Kokkos::all_libs` alias target for compatibility with TriBITS/Trilinos

### PR DESCRIPTION
**Pull request by dalg24**
Proposed alternative to #6138 

@bartlettroscoe is there a good reason `Kokkos::all_libs` cannot be an alias target?

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe is there a good reason `Kokkos::all_libs` cannot be an alias target?

@dalg24, you need both an ALIAS target and an exported IMPORTED target (the former is for users that pull in Kokkos with `add_subdirectory(kokkos)` and the latter is for users that pull in Kokkos using `find_package(Kokkos)`).

**Comment by dalg24** [comment]
> This does not provide a `Kokkos::all_libs` IMPORTED target in the generated `KokkosConfig.cmake` file (see for yourself).

I don't understand.  The configured file `KokkosConfigCommon.cmake` which gets included from `KokkosConfig.cmake` does define the alias.  What am I missing?

**Comment by masterleinad** [comment]
I can confirm that both `example/build_cmake_installed` and `example/build_cmake_in_tree` work with `Kokkos::all_libs` using this pull request.

**Review by bartlettroscoe — APPROVED** [review]
I looked over this too fast.  This adds the `Kokkos::all_libs` target to both the current CMake project and the `KokkosConfig.cmake` file.  But note this only addresses part of PR #6138.  The other part is setting missing vars like `Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE` in the `KokkosConfig.cmake` file (see https://github.com/kokkos/kokkos/pull/6138#discussion_r1197069914).

**Review by masterleinad — COMMENTED** [review]
Maybe add a comment that we do this for `Trilinos`/`TriBITS`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -325,6 +325,8 @@ IF (NOT KOKKOS_HAS_TRILINOS AND NOT Kokkos_INSTALL_TESTING)
   ADD_LIBRARY(Kokkos::kokkos ALIAS kokkos)
   TARGET_LINK_LIBRARIES(kokkos INTERFACE ${KOKKOS_COMPONENT_LIBRARIES})
   KOKKOS_INTERNAL_ADD_LIBRARY_INSTALL(kokkos)
+
+  ADD_LIBRARY(Kokkos::all_libs ALIAS kokkos)
```
Comment (masterleinad) [inline CMakeLists.txt:329]:
```suggestion
  // TriBITS requires a special target
  ADD_LIBRARY(Kokkos::all_libs ALIAS kokkos)
```

Comment (dalg24) [reply CMakeLists.txt]: I added comments

File: cmake/KokkosConfigCommon.cmake.in [hunk 2]
```diff
@@ -6,6 +6,17 @@ SET(Kokkos_CXX_COMPILER "@CMAKE_CXX_COMPILER@")
 SET(Kokkos_CXX_COMPILER_ID "@KOKKOS_CXX_COMPILER_ID@")
 SET(Kokkos_CXX_STANDARD @KOKKOS_CXX_STANDARD@)
 
+# Required to be a TriBITS-compliant external package
+IF(NOT TARGET Kokkos::all_libs)
+  # CMake Error at <prefix>/lib/cmake/Kokkos/KokkosConfigCommon.cmake:10 (ADD_LIBRARY):
+  #   ADD_LIBRARY cannot create ALIAS target "Kokkos::all_libs" because target
+  #   "Kokkos::kokkos" is imported but not globally visible.
+  IF(CMAKE_VERSION VERSION_LESS "3.18")
```
Comment (dalg24) [inline cmake/KokkosConfigCommon.cmake.in:14]: How did I determine the cmake version?
When I submitted the original version, I had tested with my local setup that has v3.25.
I was able to reproduce the CI failures with cmake-3.16 (in the CUDA RDC build that does Kokkos_INSTALL_TESTING).
After reading https://discourse.cmake.org/t/alias-and-not-globally-visible-targets/6305 I tried 3.23.5 and it did not yield an error.  Then I arbitrarily picked 3.18.0 which also worked followed by 3.17.0 that did not.

I expect this entry from the [cmake 3.18 changelog](https://cmake.org/cmake/help/latest/release/3.18.html#commands) is what resolved the issue.

> * The [add_library()](https://cmake.org/cmake/help/latest/command/add_library.html#command:add_library) and [add_executable()](https://cmake.org/cmake/help/latest/command/add_executable.html#command:add_executable) commands learned to create [Alias Targets](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#alias-targets) referencing non-GLOBAL [Imported Targets](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#imported-targets).

**Review by cz4rs — APPROVED** [review]
Looks good to me.

**Pull request merged**